src/category_theory/preadditive/injective_resolution.lean

```diff
@@ -0,0 +1,111 @@
+/-
+Copyright (c) 2022 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang, Scott Morrison
+-/
+import category_theory.preadditive.injective
+import algebra.homology.single
+
+/-!
+# Injective resolutions
+
+A injective resolution `P : InjectiveResolution Z` of an object `Z : C` consists of
+a `ℕ`-indexed cochain complex `P.cocomplex` of injective objects,
+along with a cochain map `P.ι` from cochain complex consisting just of `Z` in degree zero to `C`,
+so that the augmented chain complex is exact.
+```
+Z ----> 0 ----> ... ----> 0 ----> ...
+|       |                 |
+|       |                 |
+v       v                 v
+J⁰ ---> J¹ ---> ... ----> Jⁿ ---> ...
+```
+
+-/
+
+noncomputable theory
+
+open category_theory
+open category_theory.limits
+
+universes v u
+
+namespace category_theory
+variables {C : Type u} [category.{v} C]
+
+open injective
+
+variables [has_zero_object C] [has_zero_morphisms C] [has_equalizers C] [has_images C]
+
+/--
+An `InjectiveResolution Z` consists of a bundled `ℕ`-indexed cochain complex of injective objects,
+along with a quasi-isomorphism to the complex consisting of just `Z` supported in degree `0`.
+
+Except in situations where you want to provide a particular projective resolution
+(for example to compute a derived functor),
+you will not typically need to use this bundled object, and will instead use
+* `injective_resolution Z`: the `ℕ`-indexed cochain complex
+  (equipped with `injective` and `exact` instances)
+* `injective_resolution.ι Z`: the cochain map from  `(single C _ 0).obj Z` to
+  `injective_resolution Z` (all the components are equipped with `mono` instances,
+  and when the category is `abelian` we will show `ι` is a quasi-iso).
+-/
+@[nolint has_inhabited_instance]
+structure InjectiveResolution (Z : C) :=
+(cocomplex : cochain_complex C ℕ)
+(ι: ((cochain_complex.single₀ C).obj Z) ⟶ cocomplex)
+(injective : ∀ n, injective (cocomplex.X n) . tactic.apply_instance)
+(exact₀ : exact (ι.f 0) (cocomplex.d 0 1) . tactic.apply_instance)
+(exact : ∀ n, exact (cocomplex.d n (n+1)) (cocomplex.d (n+1) (n+2)) . tactic.apply_instance)
+(mono : mono (ι.f 0) . tactic.apply_instance)
+
+attribute [instance] InjectiveResolution.injective InjectiveResolution.exact₀
+  InjectiveResolution.exact InjectiveResolution.mono
+
+/--
+An object admits a injective resolution.
+-/
+class has_injective_resolution (Z : C) : Prop :=
+(out [] : nonempty (InjectiveResolution Z))
+
+section
+variables (C)
+
+/--
+You will rarely use this typeclass directly: it is implied by the combination
+`[enough_injectives C]` and `[abelian C]`.
+-/
+class has_injective_resolutions : Prop :=
+(out : ∀ Z : C, has_injective_resolution Z)
+
+attribute [instance, priority 100] has_injective_resolutions.out
+
+end
+
+namespace InjectiveResolution
+
+@[simp] lemma ι_f_succ {Z : C} (P : InjectiveResolution Z) (n : ℕ) :
+  P.ι.f (n+1) = 0 :=
```

Comment:
I think `P` stands for "projective". It would be good to use `I` in this file.

Reply:
done.